### ISSUE TITLE
Support custom gutter content like a button for adding a comment

### ISSUE THREAD

**Issue by kumar303** [otakustay/react-diff-view#59]
Is there a way to put something like an *add a comment* button in the gutters? Using a modified [demo site](https://github.com/otakustay/react-diff-view/tree/5d2f5590cc75d7a77d4a4e8cb6d07866d1ea355c/site) and generating a unified diff with one addition and one deletion, here's what I mean:

<img width="348" alt="Screenshot 2019-08-02 11 52 40" src="https://user-images.githubusercontent.com/55398/62386117-849b6900-b51c-11e9-852a-9d1dcbf3dac3.png">

This shows the user hovering over the line, about to press the button with a plus sign. When they press the button, some custom code could use `getWidgets()` to display a comment entry form.

How could I add a button like that?

The patch below is the only way I could think of doing it but it edits the `<GutterCell>` component within `src/Hunk/UnifiedChange.js`.

<details>
<summary>Example patch</summary>

```diff
diff --git a/site/components/DiffView/index.css b/site/components/DiffView/index.css
index e7fa0ef..c42fb32 100644
--- a/site/components/DiffView/index.css
+++ b/site/components/DiffView/index.css
@@ -25,3 +25,9 @@
     font-size: 14px;
     font-weight: bold;
 }
+
+:global(.diff-gutter-comment-button) {
+    font-size: 10px;
+    margin: 0 5px;
+    padding: 0.1em;
+}
diff --git a/site/components/DiffView/index.js b/site/components/DiffView/index.js
index f0133e6..cca057d 100644
--- a/site/components/DiffView/index.js
+++ b/site/components/DiffView/index.js
@@ -1,3 +1,4 @@
+import { useState } from 'react';
 import {
     Diff,
     Hunk,
@@ -15,6 +16,21 @@ import './index.css';
 
 const tokenize = new TokenizeWorker();
 
+const CommentButton = ({ show = false } = {}) => {
+  if (!show) {
+      return null;
+  }
+  return (
+      <button
+          className="diff-gutter-comment-button"
+          title="Add a comment"
+          type="button"
+      >
+      +
+      </button>
+  );
+};
+
 const DiffView = props => {
     const {
         hunks,
@@ -43,12 +59,32 @@ const DiffView = props => {
             : <HunkInfo key={'decoration-' + hunk.content} hunk={hunk} />;
         children.push(decorationElement);
 
+        const [showButtonForChange, setShowButtonForChange] = useState(null);
+
         const hunkElement = (
             <Hunk
                 key={'hunk-' + hunk.content}
                 hunk={hunk}
                 codeEvents={{onClick: onToggleChangeSelection}}
-                gutterEvents={{onClick: onToggleChangeSelection}}
+                gutterContent={({change}) => {
+                    return (
+                        <CommentButton
+                            show={
+                                showButtonForChange &&
+                                showButtonForChange === change
+                            }
+                        />
+                    );
+                }}
+                gutterEvents={{
+                    onClick: onToggleChangeSelection,
+                    onMouseEnter: ({change}) => {
+                        setShowButtonForChange(change);
+                    },
+                    onMouseLeave: ({change}) => {
+                        setShowButtonForChange(null);
+                    }
+                }}
             />
         );
         children.push(hunkElement);
diff --git a/src/Hunk/UnifiedHunk/UnifiedChange.js b/src/Hunk/UnifiedHunk/UnifiedChange.js
index 68f8529..da4e929 100644
--- a/src/Hunk/UnifiedHunk/UnifiedChange.js
+++ b/src/Hunk/UnifiedHunk/UnifiedChange.js
@@ -12,7 +12,16 @@ const useBoundCallbacks = (callbacks, arg) => useMemo(
     [callbacks, arg]
 );
 
-const GutterCell = ({hide, className, lineNumber, gutterAnchor, anchorID, ...props}) => {
+const GutterCell = ({
+    hide,
+    className,
+    lineNumber,
+    gutterAnchor,
+    anchorID,
+    gutterContent,
+    change,
+    ...props}
+) => {
     if (hide) {
         return null;
     }
@@ -24,6 +33,7 @@ const GutterCell = ({hide, className, lineNumber, gutterAnchor, anchorID, ...pro
                     ? <a href={'#' + anchorID} data-line-number={lineNumber}>{lineNumber}</a>
                     : lineNumber
             }
+            {gutterContent && change && gutterContent({change})}
         </td>
     );
 };
@@ -42,6 +52,7 @@ const UnifiedChange = props => {
         gutterAnchor,
         generateAnchorID,
         renderToken,
+        ...gutterProps
     } = props;
     const {type, content} = change;
     const oldLine = computeOldLineNumber(change);
@@ -78,7 +89,9 @@ const UnifiedChange = props => {
                 lineNumber={oldLineNumber}
                 gutterAnchor={gutterAnchor}
                 anchorID={anchorID}
+                change={change}
                 {...boundGutterEvents}
+                {...gutterProps}
             />
             <GutterCell
                 hide={hideGutter}

```

</details>

This patch was made against https://github.com/otakustay/react-diff-view/commit/cc77b3248c6ee76a5da048985561d24e8905cab8 and does the following:

- Exploits the `gutterEvents` object by adding `onMouseEnter` and `onMouseLeave` to detect when the user is hovering over a line. ~~That's unsupported and fragile.~~ (I re-read the docs; it's supported.)
- Adds a new prop to `Hunk` called `gutterContent` which is a callback taking a `change` object and returning JSX to put in the gutter. This would need to be a new feature.

**Comment by kumar303** [comment]
Oh, I thought of a way to do it without modifying the lib.

<details>
<summary>Example patch to the demo site</summary>

```diff
diff --git a/site/components/DiffView/index.css b/site/components/DiffView/index.css
index e7fa0ef..6da80fd 100644
--- a/site/components/DiffView/index.css
+++ b/site/components/DiffView/index.css
@@ -25,3 +25,19 @@
     font-size: 14px;
     font-weight: bold;
 }
+
+:global(.diff-widget) {
+    position: relative;
+}
+
+:global(.diff-gutter-comment-button) {
+    align-items: center;
+    bottom: -12px;
+    display: flex;
+    font-size: 12px;
+    height: 24px;
+    justify-content: center;
+    left: 12px;
+    position: absolute;
+    width: 24px;
+}
diff --git a/site/components/DiffView/index.js b/site/components/DiffView/index.js
index f0133e6..60d10a6 100644
--- a/site/components/DiffView/index.js
+++ b/site/components/DiffView/index.js
@@ -1,6 +1,8 @@
+import { useState } from 'react';
 import {
     Diff,
     Hunk,
+    getChangeKey,
     withSourceExpansion,
     minCollapsedLines,
     withChangeSelect,
@@ -15,6 +17,23 @@ import './index.css';
 
 const tokenize = new TokenizeWorker();
 
+const CommentButton = ({
+    change,
+    setShowButtonForChange,
+} = {}) => {
+    return (
+        <button
+            className="diff-gutter-comment-button"
+            onMouseEnter={() => setShowButtonForChange(change)}
+            onMouseLeave={() => setShowButtonForChange(null)}
+            title="Add a comment"
+            type="button"
+        >
+        +
+        </button>
+    );
+};
+
 const DiffView = props => {
     const {
         hunks,
@@ -43,12 +62,44 @@ const DiffView = props => {
             : <HunkInfo key={'decoration-' + hunk.content} hunk={hunk} />;
         children.push(decorationElement);
 
+        const [showButtonForChange, setShowButtonForChange] = useState(null);
+
+        // Show an 'add comment' button any time one hovers over a line
+        // of code or in the gutter.
+        const showButtonEvents = {
+            onMouseEnter: ({change}) => {
+                setShowButtonForChange(change);
+            },
+            onMouseLeave: ({change}) => {
+                setShowButtonForChange(null);
+            }
+        };
+
+        const hunkProps = {};
+        if (showButtonForChange) {
+            hunkProps.widgets = {
+                [getChangeKey(showButtonForChange)]: (
+                    <CommentButton
+                        change={showButtonForChange}
+                        setShowButtonForChange={setShowButtonForChange}
+                    />
+                )
+            };
+        }
+
         const hunkElement = (
             <Hunk
                 key={'hunk-' + hunk.content}
                 hunk={hunk}
-                codeEvents={{onClick: onToggleChangeSelection}}
-                gutterEvents={{onClick: onToggleChangeSelection}}
+                codeEvents={{
+                    onClick: onToggleChangeSelection,
+                    ...showButtonEvents
+                }}
+                gutterEvents={{
+                    onClick: onToggleChangeSelection,
+                    ...showButtonEvents
+                }}
+                {...hunkProps}
             />
         );
         children.push(hunkElement);

```

</details>

The button is rendered with `getWidgets()`:

<img width="346" alt="Screenshot 2019-08-02 15 25 10" src="https://user-images.githubusercontent.com/55398/62397320-19ad5a80-b53b-11e9-966b-61f3a88e4e67.png">


Well, I guess that will work! Feel free to close the issue but let me know if there's an easier way.

**Comment by otakustay** [comment]
A widget with offset doesn't seem to be a good idea, I'd more like to add a `renderGutter` prop so that we can have a more complex component in gutter cell

This prop can also help to mark unit test coverage info on gutter edge like a red or green bar considering #48 

Would you like to evaluate if this could work:

1. `Hunk` component will add a `renderGutter({change, side})` prop with default behavior to render the line number
2. Use `onMouseEnter` and `onMouseLeave` and a local state to determine which line is currently hovered
2. Customize `renderGutter` to render either line number or command button determined by currently hovered line

This may more or less like:

```jsx
const [hoveredChange, setHoveredChange] = useState(null);
const renderGutter = ({change, side}) => {
    if (change === hoveredChange) {
        return <CommandButton />;
    }

    return side === 'old' ? computeOldLineNumber(change): computeNewLineNumber(change);
};
const customEvents = useMemo(
    () => {
        return {
            onMouseEnter({change}) {
                setHoveredChange(change);
            },
        };
    },
    []
);
const renderHunk = hunk => (
    <Hunk
        key={hunk.content}
        hunk={hunk}
        renderGutter={renderGutter}
        customEvents={customEvents}
    />
);

<Diff>
    {hunks => hunks.map(renderHunk)}
</Diff>
```

I'll have a try of this pattern with consideration of performance in a large diff later.

**Comment by otakustay** [comment]
By the way, we simply add a comment by double clicking any change in our team...

**Comment by otakustay** [comment]
@kumar303 We have roughly a version of `renderGutter` which is quite performance friendly since I tested it against a diff with 12890 changes.

To ensure performance, we cannot handle hover state outside `Diff` component, it could leads every cell to re-render on mouse moves, so the final proposal could be:

```jsx
const renderGutter = ({change, side, hover}) => {
    if (hover) {
        return (
            <Tooltip title="Not implemented yet">
                <span className={c.commentTrigger}>+</span>
            </Tooltip>
        );
    }

    const lineNumber = side === 'new' ? computeNewLineNumber(change) : computeOldLineNumber(change);
    return lineNumber === -1 ? '' : lineNumber.toString();
};

<Diff renderGutter={renderGutter}>
    ...
</Diff>
```

We provide these properties to `renderGutter`'s single argument:

- `{Change} change`: current change.
- `{string} side`: `"new"` or `"old"`.
- `{boolean} gutterAnchor`: Same prop on `Diff` component
- `{string} anchorID`: Generated anchor target id, custom implement should render a `<a href={`#${anchorID}`}` when `gutterAnchor` is `true`.
- `{boolean} hover`: whether this change is hovered.

I wonder if a boolean `hover` property is the adequate name to tell the hover state, any idea?

**Comment by otakustay** [comment]
https://otakustay.github.io/react-diff-view/ is updated to demonstrated custom `renderGutter` prop

**Comment by kumar303** [comment]
I like the custom `renderGutter()` idea. Here are some thoughts:

* A custom `renderGutter()` shouldn't be responsible for rendering the line number. Instead, `renderGutter()` could return something like `{ content: <CustomStuff />, replace: false }` where `replace` would indicate if it should replace the line number rendering or simply add to it.
* I agree that `onMouseEnter` is not great for performance. I'd rather use CSS for this which, now that I've thought more about it, should totally be possible.

Here is another prototype that implements some of what I mentioned above.

<details>
<summary>demo</summary>

![2019-08-08 15 06 38](https://user-images.githubusercontent.com/55398/62734624-0f81d500-b9ef-11e9-965d-8f2335a8dd59.gif)

</details>


<details>
<summary>patch</summary>

```diff
diff --git a/site/components/DiffView/index.css b/site/components/DiffView/index.css
index e7fa0ef..fc31c00 100644
--- a/site/components/DiffView/index.css
+++ b/site/components/DiffView/index.css
@@ -25,3 +25,14 @@
     font-size: 14px;
     font-weight: bold;
 }
+
+:global(.diff-gutter-comment-button) {
+    cursor: pointer;
+    font-size: 10px;
+    margin: 0 5px;
+    padding: 0.1em;
+}
+
+:global(.diff-line:not(:hover) .diff-gutter-comment-button) {
+    visibility: hidden;
+}
diff --git a/site/components/DiffView/index.js b/site/components/DiffView/index.js
index f0133e6..360787f 100644
--- a/site/components/DiffView/index.js
+++ b/site/components/DiffView/index.js
@@ -1,3 +1,4 @@
+import { useState } from 'react';
 import {
     Diff,
     Hunk,
@@ -15,6 +16,18 @@ import './index.css';
 
 const tokenize = new TokenizeWorker();
 
+const CommentButton = () => {
+  return (
+      <button
+          className="diff-gutter-comment-button"
+          title="Add a comment"
+          type="button"
+      >
+      +
+      </button>
+  );
+};
+
 const DiffView = props => {
     const {
         hunks,
@@ -48,7 +61,12 @@ const DiffView = props => {
                 key={'hunk-' + hunk.content}
                 hunk={hunk}
                 codeEvents={{onClick: onToggleChangeSelection}}
-                gutterEvents={{onClick: onToggleChangeSelection}}
+                renderGutter={({change, side}) => {
+                    return {
+                        content: side === 'new' && <CommentButton />,
+                        replace: false,
+                    };
+                }}
             />
         );
         children.push(hunkElement);
diff --git a/src/Hunk/UnifiedHunk/UnifiedChange.js b/src/Hunk/UnifiedHunk/UnifiedChange.js
index c67239a..aae241e 100644
--- a/src/Hunk/UnifiedHunk/UnifiedChange.js
+++ b/src/Hunk/UnifiedHunk/UnifiedChange.js
@@ -12,15 +12,26 @@ const useBoundCallbacks = (callbacks, arg) => useMemo(
     [callbacks, arg]
 );
 
-const renderGutterCell = (className, lineNumber, gutterAnchor, anchorID, events) => (
-    <td className={className} {...events}>
-        {
-            gutterAnchor
-                ? <a href={'#' + anchorID} data-line-number={lineNumber}>{lineNumber}</a>
-                : lineNumber
-        }
-    </td>
-);
+const renderGutterCell = (className, lineNumber, gutterAnchor, anchorID, events, customGutterInfo) => {
+    let useCustomContentOnly = false;
+    let customContent;
+
+    if (customGutterInfo) {
+        customContent = customGutterInfo.content;
+        useCustomContentOnly = customGutterInfo.replace;
+    }
+
+    return (
+        <td className={className} {...events}>
+            {
+                !useCustomContentOnly && gutterAnchor
+                    ? <a href={'#' + anchorID} data-line-number={lineNumber}>{lineNumber}</a>
+                    : lineNumber
+            }
+            {customContent}
+        </td>
+    );
+};
 
 const UnifiedChange = props => {
     const {
@@ -36,6 +47,7 @@ const UnifiedChange = props => {
         gutterAnchor,
         generateAnchorID,
         renderToken,
+        renderGutter,
     } = props;
     const {type, content} = change;
     const oldLine = computeOldLineNumber(change);
@@ -61,6 +73,12 @@ const UnifiedChange = props => {
         {'diff-code-selected': selected}
     );
 
+    let customGutterInfo;
+    if (renderGutter) {
+        // TODO: make one for each side.
+        customGutterInfo = renderGutter({ change });
+    }
+
     return (
         <tr id={anchorID} className={classNames('diff-line', className)}>
             {
@@ -69,7 +87,8 @@ const UnifiedChange = props => {
                     oldLineNumber,
                     gutterAnchor,
                     anchorID,
-                    boundGutterEvents
+                    boundGutterEvents,
+                    renderGutter && renderGutter({ change, side: 'old' }),
                 )
             }
             {
@@ -78,7 +97,8 @@ const UnifiedChange = props => {
                     newLineNumber,
                     gutterAnchor,
                     anchorID,
-                    boundGutterEvents
+                    boundGutterEvents,
+                    renderGutter && renderGutter({ change, side: 'new' }),
                 )
             }
             <CodeCell

```

</details>

I don't know if it would cover all use cases but this would cover what I need, for sure.

**Comment by otakustay** [comment]
`{replace: false}` is a bit ambiguous since

1. we can't tell we should insert custom content before or after line numbers, or even have multiple elements surrounding line numbers.
2. `gutterAnchor` is an indication that you should wrap all content in an anchor element with a href to `#${anchorID}` so it gets url hash updated, but it's not a must, we're not confident to automatically put every element `renderGutter` returned into a `<a>`.

I'd propose a signature of `renderGutter({change, side, hover}, renderDefault, wrapAnchor)` to help developers work like:

```js
const renderGutter = ({change, side, hover}, renderDefault, wrapAnchor) => {
  const content = (
    <>
      <UnitTestCoverageBar covered={covered(change)} />
      {renderDefault()}
      <CommandButton />
    </>
  );
  return wrapAnchor(content);
};
```

The CSS `:hover` selector works with unified view, but will fail in split view since it is unable to determine which side is hovered, this is the main reason we provide an internal hover implementation via `onMouseEnter` and `onMouseLeave`

**Comment by kumar303** [comment]
I like the `renderDefault()` idea! That's nice.

As for your question about naming the `hover` parameter, maybe call it `inHoverState` or `inHover`? This would convey that it is a flag to indicate the state.

I have a question about the value of `side` (`new` or `old`). How would that work for `normal` changes? I guess those would have `side: 'old'`? I wonder if one would need to differentiate normal changes in `renderGutter()`.

Also, to let you evolve the callback signature more easily in future releases, I suggest using a single object for all parameters. Taking your last comment as an example, that would look like:

```js
const renderGutter = ({change, side, hover, wrapAnchor, renderDefault}) => {
}
```

A minor point: `warpAnchor` would read better to me as `wrapInAnchor`.

**Comment by otakustay** [comment]
> How would that work for `normal` changes?

Split view is quite straightforward:

![image](https://user-images.githubusercontent.com/639549/62784691-a20a9d00-baf1-11e9-99e3-940fc82139ec.png)

For unified view:

![image](https://user-images.githubusercontent.com/639549/62784920-1a715e00-baf2-11e9-8168-bdd509d90f89.png)

The 2 empty gutter cells will not invoke `renderGutter`

> I suggest using a single object for all parameters

Great, I'll work this way, `wrapInAnchor` and `inHoverState` will be our final name

**Comment by kumar303** [comment]
re: normal changes, I see now. That will work.

Thanks for your quick response on this!

**Comment by otakustay** [comment]
Actually I'm now wondering about the "2 empty gutters will not invoke `renderGutter`" statement, is it possible that we need something like this:

![image](https://user-images.githubusercontent.com/639549/62791094-0c760a00-baff-11e9-864d-ce68a9a33119.png)

A command button that always resides in the second column, even on the deletion row, in this case we have to invoke `renderGutter` for all cells, leave determination to developers

Currently I'd prefer to invoke `renderGutter` on all gutter cells no matter it should be empty or not

**Comment by kumar303** [comment]
> Currently I'd prefer to invoke `renderGutter` on all gutter cells no matter it should be empty or not

Yeah, I think that would be the best.

For something like a comment button, the UI would be easier to use if the button was always in the same place.

**Comment by otakustay** [comment]
We published `2.2.0` to add `renderGutter` prop, https://otakustay.github.io/react-diff-view/ is updated, feedbacks are welcome at any time so we can have quick fixes to related issues.
